src/com/ka/spreadsheet/diff/UnifiedDiffSpreadSheetDiffCallback.java

```diff
@@ -0,0 +1,168 @@
+    assert previousCell == null || c.compareCellPositions(previousCell) >= 0 :
+      "Cell-ordering contract violated.  Previous=" + previousCell.getCellPosition()
+      + ", current=" + c.getCellPosition();
+    previousCell = c;
+    accumulateAndMaybePrint(new DiffCell(
+      c.getSheetName(),
+      c.getRowIndex(),
+      c.getColumnIndex(),
+      (inFirstSpreadSheet ? c.getCellValue() : null),
+      (!inFirstSpreadSheet ? c.getCellValue() : null)
+    ));
+  }
+
+  @Override
+  public void reportDiffCell(CellPos c1, CellPos c2) {
+    assert (c1.getRowIndex() == c2.getRowIndex())
+      && (c1.getColumnIndex() == c2.getColumnIndex()) : "Cells are not at the same position. Cell 1="
+      + c1.getCellPosition() + ", cell 2=" + c2.getCellPosition();
+    assert previousCell == null || c1.compareCellPositions(previousCell) >= 0 :
+      "Cell-ordering contract violated.  Previous=" + previousCell.getCellPosition()
+      + ", current=" + c1.getCellPosition();
+    previousCell = c1;
+    accumulateAndMaybePrint(new DiffCell(
+      c1.getSheetName(),
+      c1.getRowIndex(),
+      c1.getColumnIndex(),
+      c1.getCellValue(),
+      c2.getCellValue()
+    ));
+  }
+
+  private void accumulateAndMaybePrint(DiffCell diffCell) {
+    if (!isSameSheet(prevDiffCell, diffCell)) {
+      printAndEmptyCellBlock();
+      System.out.println("--- " + file1 + "!" + diffCell.sheetName);
+      System.out.println("+++ " + file2 + "!" + diffCell.sheetName);
+    }
+    if (!isSameRow(prevDiffCell, diffCell)) {
+      printAndEmptyCellBlock();
+    }
+    if (!isSameCellBlock(prevDiffCell, diffCell)) {
+      printAndEmptyCellBlock();
+    }
+    currentCellBlock.add(diffCell);
+    prevDiffCell = diffCell;
+  }
+
+  private boolean isSameSheet(DiffCell c1, DiffCell c2) {
+    assert c1 != null && c2 != null;
+    return c1.sheetName.equals(c2.sheetName);
+  }
+
+  private boolean isSameRow(DiffCell c1, DiffCell c2) {
+    assert c1 != null && c2 != null;
+    return c1.rowIndex == c2.rowIndex;
+  }
+
+  private boolean isSameCellBlock(DiffCell c1, DiffCell c2) {
+    assert c1 != null && c2 != null;
+    return c1.colIndex == (c2.colIndex - 1);
+  }
+
+  // TODO: Make this handle multiple rows, maybe.  What would that output look like?  This might be a bad idea.
+  private void printAndEmptyCellBlock() {
+    assert currentCellBlock != null;
+    if (currentCellBlock.size() > 0) {
+      StringBuilder sheet1Lines = new StringBuilder();
+      StringBuilder sheet2Lines = new StringBuilder();
+      String cellRange = CELL_INTERNAL_TO_USER(currentCellBlock.get(0).rowIndex, currentCellBlock.get(0).colIndex);
+      if (currentCellBlock.size() > 1) {
+          cellRange = cellRange + "," +
+            CELL_INTERNAL_TO_USER(currentCellBlock.get(0).rowIndex, currentCellBlock.get(currentCellBlock.size()-1).colIndex);
+      }
+      System.out.println("@@ -" + cellRange + " +" + cellRange + " @@");
```

Comment:
you're overwriting the cellRange variable, you need startCellRange, endCellRange, or print early

Reply:
That's intentional.  I wanted `cellRange` to be either a single cell or a range of cells (_e.g._, `A1` or `A1,C1`), and the "@@" line contains the old-file and new-file ranges.  Due to the way the current diff engine works, the old and new ranges will always be the same, hence this construction.  Thus `println()` will always output `@@ -old_range +new_range @@`, where a range is either a single cell (_e.g._, `@@ -A1 +A1 @@`) or a starting and ending cell id (_e.g._, `@@ -A1,C1 +A1,C1 @@`).

However, looking at that, I realize I translated from the output format of `diff -u` incorrectly.  The range format is a hold-over from my original thought of supporting 2D cell blocks, but I figured out pretty early in this effort that 1D was the best I could do.

What `diff` actually outputs for a range isn't `start_line,end_line`, it's `start_line,line_count`, which makes sense, because `diff` is line-oriented.  I could change to having a cell range be a starting cell id and a cell count (_e.g._, `@@ -A1,3 +A1,3 @@`), but I actually think the current not-quite-diff-u ranges are clearer.  And unless there's a diff engine that can identify insertions, the "old" and "new" ranges will always be the same, which isn't the case in `diff`.

Do you have a preference?  I'm leaning towards keeping the current format (in which case, this code is correct).

Reply:
Keeping the current format is good